sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala

```diff
@@ -268,6 +268,27 @@ private[sql] abstract class SparkStrategies extends QueryPlanner[SparkPlan] {
 
   object CartesianProduct extends Strategy {
     def apply(plan: LogicalPlan): Seq[SparkPlan] = plan match {
+      // Not like the equal-join, BroadcastNestedLoopJoin doesn't support condition
+      // for cartesian join, as in cartesian join, probably, the records satisfy the
+      // condition, but exists in another partition of the large table, so we may not able
+      // to eliminate the duplicates.
+      case logical.Join(
+        CanBroadcast(left), right, joinType @ (FullOuter | LeftOuter | RightOuter), None) =>
+        execution.joins.BroadcastNestedLoopJoin(
+          planLater(left), planLater(right), joins.BuildLeft, joinType, None) :: Nil
```

Comment:
I may miss some context at here. This is a strategy for CartesianProduct. We have a strategy `BroadcastNestedLoopJoin` to handle other cases.